File: system-test/dotnet/SystemTests/Legacy/Mod03/E2eTests/PlaceOrderPositiveUiTest.cs
using System.Text;
using System.Text.RegularExpressions;
using SystemTests.Commons.Constants;
using SystemTests.Legacy.Mod03.E2eTests.Base;
using Shouldly;
using Xunit;

namespace SystemTests.Legacy.Mod03.E2eTests;

public class PlaceOrderPositiveUiTest : BaseE2eTest
{
    protected override async Task SetShopRawAsync()
    {
        await SetUpShopBrowserAsync();
    }

    [Fact]
    public async Task ShouldPlaceOrderForValidInput()
    {
        var sku = CreateUniqueSku(Defaults.SKU);
        var createProductJson = $$"""{"id":"{{sku}}","title":"Test Product","description":"Test Description","category":"Test Category","brand":"Test Brand","price":"20.00"}""";

        var createProductUri = new Uri(_configuration.ErpBaseUrl + "/api/products");
        var createProductContent = new StringContent(createProductJson, Encoding.UTF8, "application/json");
        var createProductResponse = await _erpHttpClient!.PostAsync(createProductUri, createProductContent);
        ((int)createProductResponse.StatusCode).ShouldBe(201);

        await shopUiPage!.GotoAsync(_configuration.ShopUiBaseUrl);
        await shopUiPage.Locator("a[href='/new-order']").ClickAsync();

        await shopUiPage.Locator("[aria-label=\"SKU\"]").FillAsync(sku);
        await shopUiPage.Locator("[aria-label=\"Quantity\"]").FillAsync("5");
        await shopUiPage.Locator("[aria-label=\"Place Order\"]").ClickAsync();

        var successMessageText = await shopUiPage.Locator("[role='alert']").TextContentAsync();
        var match = Regex.Match(successMessageText ?? "", @"Success! Order has been created with Order Number ([\w-]+)");
        match.Success.ShouldBeTrue();
        var orderNumber = match.Groups[1].Value;
        orderNumber.ShouldStartWith("ORD-");

        await shopUiPage.GotoAsync(_configuration.ShopUiBaseUrl);
        await shopUiPage.Locator("a[href='/order-history']").ClickAsync();
        await shopUiPage.Locator("[aria-label='Order Number']").FillAsync(orderNumber);
        await shopUiPage.Locator("[aria-label='Refresh Order List']").ClickAsync();

        var rowSelector = $"xpath=//tr[contains(., '{orderNumber}')]";
        await shopUiPage.Locator(rowSelector).WaitForAsync(new() { State = Microsoft.Playwright.WaitForSelectorState.Visible });
        (await shopUiPage.Locator(rowSelector).IsVisibleAsync()).ShouldBeTrue();

        var viewDetailsSelector = $"xpath=//tr[contains(., '{orderNumber}')]//a[contains(text(), 'View Details')]";
        await shopUiPage.Locator(viewDetailsSelector).ClickAsync();

        (await shopUiPage.Locator("[aria-label='Display Order Number']").TextContentAsync()).ShouldBe(orderNumber);
        (await shopUiPage.Locator("[aria-label='Display SKU']").TextContentAsync()).ShouldBe(sku);
        int.Parse(await shopUiPage.Locator("[aria-label='Display Quantity']").TextContentAsync() ?? "0").ShouldBe(5);

        var unitPriceText = await shopUiPage.Locator("[aria-label='Display Unit Price']").TextContentAsync();
        decimal.Parse((unitPriceText ?? "").Replace("$", "")).ShouldBe(20.00m);

        var totalPriceText = await shopUiPage.Locator("[aria-label='Display Total Price']").TextContentAsync();
        decimal.Parse((totalPriceText ?? "").Replace("$", "")).ShouldBeGreaterThan(0);

        (await shopUiPage.Locator("[aria-label='Display Status']").TextContentAsync()).ShouldBe("PLACED");
    }
}

File: system-test/java/src/test/java/com/optivem/shop/systemtest/legacy/mod03/e2e/PlaceOrderPositiveUiTest.java
package com.optivem.shop.systemtest.legacy.mod03.e2e;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.options.WaitForSelectorState;
import com.optivem.shop.systemtest.legacy.mod03.e2e.base.BaseE2eTest;
import org.junit.jupiter.api.Test;

import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.UUID;
import java.util.regex.Pattern;

import static com.optivem.shop.systemtest.commons.constants.Defaults.*;
import static org.assertj.core.api.Assertions.assertThat;

class PlaceOrderPositiveUiTest extends BaseE2eTest {
    @Override
    protected void setShopClient() {
        setUpShopBrowser();
    }

    @Test
    void shouldPlaceOrderForValidInput() throws Exception {
        var sku = SKU + "-" + UUID.randomUUID().toString().substring(0, 8);
        var createProductJson = """
                {
                    "id": "%s",
                    "title": "Test Product",
                    "description": "Test Description",
                    "category": "Test Category",
                    "brand": "Test Brand",
                    "price": "20.00"
                }
                """.formatted(sku);

        var createProductUri = URI.create(getErpBaseUrl() + "/api/products");
        var createProductRequest = HttpRequest.newBuilder()
                .uri(createProductUri)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(createProductJson))
                .build();

        var createProductResponse = erpHttpClient.send(createProductRequest, HttpResponse.BodyHandlers.ofString());
        assertThat(createProductResponse.statusCode()).isEqualTo(201);

        shopUiPage.navigate(getShopUiBaseUrl());
        shopUiPage.locator("a[href='/new-order']").click();

        shopUiPage.locator("[aria-label=\"SKU\"]").fill(sku);
        shopUiPage.locator("[aria-label=\"Quantity\"]").fill("5");
        shopUiPage.locator("[aria-label=\"Country\"]").fill(COUNTRY);
        shopUiPage.locator("[aria-label=\"Place Order\"]").click();

        var successMessageText = shopUiPage.locator("[role='alert']").textContent();
        var pattern = Pattern.compile("Success! Order has been created with Order Number ([\\w-]+)");
        var matcher = pattern.matcher(successMessageText);
        assertThat(matcher.find()).isTrue();
        var orderNumber = matcher.group(1);
        assertThat(orderNumber).startsWith("ORD-");

        shopUiPage.navigate(getShopUiBaseUrl());
        shopUiPage.locator("a[href='/order-history']").click();
        shopUiPage.locator("[aria-label='Order Number']").fill(orderNumber);
        shopUiPage.locator("[aria-label='Refresh Order List']").click();

        var rowSelector = String.format("//tr[contains(., '%s')]", orderNumber);
        shopUiPage.locator(rowSelector).waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE));
        assertThat(shopUiPage.locator(rowSelector).isVisible()).isTrue();

        var viewDetailsSelector = String.format("%s//a[contains(text(), 'View Details')]", rowSelector);
        shopUiPage.locator(viewDetailsSelector).click();

        shopUiPage.locator("[aria-label='Display Order Number']")
                .waitFor(new Locator.WaitForOptions().setState(WaitForSelectorState.VISIBLE));
        assertThat(shopUiPage.locator("[aria-label='Display Order Number']").textContent()).isEqualTo(orderNumber);
        assertThat(shopUiPage.locator("[aria-label='Display SKU']").textContent()).isEqualTo(sku);
        assertThat(Integer.parseInt(shopUiPage.locator("[aria-label='Display Quantity']").textContent())).isEqualTo(5);

        var unitPriceText = shopUiPage.locator("[aria-label='Display Unit Price']").textContent().replace("$", "");
        assertThat(Double.parseDouble(unitPriceText)).isEqualTo(20.00);

        var totalPriceText = shopUiPage.locator("[aria-label='Display Total Price']").textContent().replace("$", "");
        assertThat(Double.parseDouble(totalPriceText)).isGreaterThan(0.0);

        assertThat(shopUiPage.locator("[aria-label='Display Status']").textContent()).isEqualTo("PLACED");
    }
}
